feat: Support for encrypting the database and bucket with CMK

velotioaastha opened the pull request:
The supported scenarios are listed in this doc - 
https://docs.google.com/document/d/1DAQQx2h1weTDCsLNuH0jxJwfJpm95AIQgfCedIQkYlU/edit?usp=sharing

gls4 reviewed (changes requested):
So we're creating a KMS key by default for use with S3 and RDS. I'd like to get rid of the boolean `create_kms_key` and use the default key _unless_ the customer provides us the ARN of a key to use for S3.  The same for the database. This will allow for simpler (cleaner) code in several places, namely where we've terniary operators. 

Let's amend the name of the default KMS key to `kms_default_key` instead of using the word `generic` in the name of variables and modules. Let me know if you have questions and/or would like to discuss.

gls4 commented on main.tf:
```diff
@@ -8,7 +8,13 @@ module "kms" {
 locals {
-  kms_key_arn         = module.kms.key.arn
+
+  kms_key_arn_generic = module.kms.key.arn
```
Let's change this to `default_kms_key`.

gls4 commented on main.tf:
```diff
@@ -8,7 +8,13 @@ module "kms" {
+
+  kms_key_arn_generic = module.kms.key.arn
+
+  s3_kms_key_arn= var.create_kms ? local.kms_key_arn_generic : length(var.bucket_kms_key_arn)> 0 ? var.bucket_kms_key_arn : local.kms_key_arn_generic
```
I'd like this statement to be a bit less complex. Let's check for and use the customer-provided KMS arn -OR- use the ARN of the default key.

gls4 replied:
So let's do this:

1. Always create the default keyys
2. Use the default keys unless the customer provides them
3. Let's validate the KMS ARN on the customer provided value, so we accept either `null` or a valid ARN.

gls4 replied:
We need to test how validation behaves -- I'm sure there's a way to test for a valid KMS ARN -OR- null, but we need to be sure. That said, I would use a `validation` block something like this:
```
validation {
    # regex(...) fails if it cannot find a match
    condition     = can(regex("^arn:aws:kms:*:${Account}:*/${Id}"))
    error_message = "..."
  }
  ```

velotioaastha replied:
Hi @gls4,
okay, we can handle this condition with kms arn or null within the condition itself.
One more thing, Should we not assign default key in case create_kms is false and kms_key_arn is null.

gls4 commented on main.tf:
```diff
@@ -8,7 +8,13 @@ module "kms" {
+
+  s3_kms_key_arn= var.create_kms ? local.kms_key_arn_generic : length(var.bucket_kms_key_arn)> 0 ? var.bucket_kms_key_arn : local.kms_key_arn_generic
+  
+  db_kms_key_arn = var.create_kms ?  local.kms_key_arn_generic: length(var.db_kms_key_arn)> 0  ? var.db_kms_key_arn : local.kms_key_arn_generic
```
Same thing here:  check for and use the customer-provided KMS arn -OR- use the ARN of the default key.

gls4 replied:
I like what you've suggested. Having `create_kms` seems redundant, since we're always going to create at least one.

velotioaastha replied:
hi @gls4,

okay, so you mean we should simply put this condition only.. 

s3_kms_key_arn= length(var.bucket_kms_key_arn)> 0 ? var.bucket_kms_key_arn : local.default_kms_key

db_kms_key_arn = length(var.db_kms_key_arn)> 0  ? var.db_kms_key_arn : local.default_kms_key

With this, we don't need to create another boolean which was telling us whether to encrypt s3 or not.
Knowing that null will always result in attaching our generic keys. and the bucket will always be encrypted.

gls4 replied:
Yes, that's it! :-)

gls4 commented on main.tf:
```diff
@@ -61,7 +67,7 @@ module "database" {
 
   namespace                        = var.namespace
-  kms_key_arn                      = local.kms_key_arn
+  kms_key_arn                      = local.db_kms_key_arn
```
Can we use the default KMS key/customer-provided KMS key here as well?

velotioaastha replied:
hey George, 

Actually, this the local variable, is dependent on conditions and separate variables. That's why we named it differently.

This is the place where this variable value is [determined](https://github.com/wandb/terraform-aws-wandb/blob/aastha/encryption/main.tf#L16).

gls4 replied:
Ah, yes. Thank you for clarifying.

gls4 commented on main.tf:
```diff
@@ -117,15 +123,15 @@ module "app_eks" {
 
   namespace   = var.namespace
-  kms_key_arn = local.kms_key_arn
+  kms_key_arn = local.kms_key_arn_generic 
```
Let's change this to`default_kms_key_arn`

gls4 commented on outputs.tf:
```diff
@@ -48,7 +48,7 @@ output "internal_app_port" {
 
 output "kms_key_arn" {
-  value       = local.kms_key_arn
+  value       = local.kms_key_arn_generic
```
Change the name to `default_kms_key_arn`

gls4 commented on variables.tf:
```diff
@@ -392,3 +393,18 @@ variable "other_wandb_env" {
+  default = true
+
+}
+variable "db_kms_key_arn" {
```
Change the name in keeping with previous comments.

gls4 reviewed (changes requested):
A few changes requested: variable names, consideration of performance insights key, remove boolean which toggles creation of KMS key.

velotioaastha commented:
Hey @gls4, We have added changes for performance insights kms key and updated the naming convention for default kms key. 

Can you also suggest on kms ARN conditions. I've added as a comment there. Will update accordingly.
Please review the PR.

gls4 commented:
I need to test this. More to come.

gls4 commented:
Hello! Looks like we might've missed changing a variable name here:
```
 Error: Reference to undeclared local value
│ 
│   on ../../outputs.tf line 51, in output "kms_key_arn":
│   51:   value       = local.kms_key_arn_generic
│ 
│ A local value with the name "kms_key_arn_generic" has not been declared.
```

velotioaastha commented:
hey @gls4, all the comments are resolved. And the PR can be reviewed again.

jsbroks commented:
This PR is included in version 4.20.0 :tada:
